File: go/vectors_test.go
package fiki

// The shared conformance vectors (`this.i` @5gf6r08f, @2tt6fmc0).
//
// They live at the repository root rather than under go/ so this implementation and the other four
// are held to the same bytes. A copy under each language is the drift the polyglot layout exists
// to prevent, which is why this file reaches up two directories rather than embedding anything.

import (
	"encoding/base64"
	"encoding/hex"
	"encoding/json"
	"errors"
	"os"
	"path/filepath"
	"slices"
	"testing"
)

func load(t *testing.T, name string, into any) {
	t.Helper()
	raw, err := os.ReadFile(filepath.Join("..", "vectors", name))
	if err != nil {
		t.Fatalf("the shared vectors are not where every port reaches them: %v", err)
	}
	if err := json.Unmarshal(raw, into); err != nil {
		t.Fatalf("%s: %v", name, err)
	}
}

type aidCase struct {
	ID           string `json:"id"`
	SeedHex      string `json:"seed_hex"`
	PublicKeyHex string `json:"public_key_hex"`
	AID          string `json:"aid"`
	Keyid        string `json:"keyid"`
}

type baseCase struct {
	ID        string            `json:"id"`
	SeedHex   string            `json:"seed_hex"`
	Method    string            `json:"method"`
	URL       string            `json:"url"`
	Headers   map[string]string `json:"headers"`
	Covered   []string          `json:"covered"`
	Created   int64             `json:"created"`
	Keyid     string            `json:"keyid"`
	Alg       string            `json:"alg"`
	Base      string            `json:"base"`
	Signature string            `json:"signature"`
}

type requestCase struct {
	ID      string            `json:"id"`
	Method  string            `json:"method"`
	URL     string            `json:"url"`
	Headers map[string]string `json:"headers"`
	Body    *string           `json:"body"`
	MaxAge  *int64            `json:"max_age"`
	Now     *int64            `json:"now"`
	AID     string            `json:"aid"`
	Covered []string          `json:"covered"`
	Error   string            `json:"error"`
}

func (c requestCase) body() []byte {
	if c.Body == nil {
		return nil
	}
	return []byte(*c.Body)
}

func (c requestCase) options() VerifyOptions {
	opts := VerifyOptions{MaxAge: c.MaxAge, Body: c.body()}
	if c.Now != nil {
		opts.Now = *c.Now
	}
	return opts
}

func mustHex(t *testing.T, text string) []byte {
	t.Helper()
	raw, err := hex.DecodeString(text)
	if err != nil {
		t.Fatal(err)
	}
	return raw
}

func TestThisPortSatisfiesTheVectorsFormatItIsRunning(t *testing.T) {
	// A port running newer vectors fails here rather than passing a subset and reporting
	// conformance it no longer has: the cases it never implemented would simply not be in the
	// file it last read.
	for _, name := range []string{"aid-lens.json", "signature-base.json", "accepts.json", "refusals.json"} {
		t.Run(name, func(t *testing.T) {
			var file struct {
				VectorsFormat int `json:"vectors_format"`
			}
			load(t, name, &file)
			if file.VectorsFormat != VectorsFormat {
				t.Errorf("%s declares vectors format %d; this port satisfies %d", name, file.VectorsFormat, VectorsFormat)
			}
		})
	}
}

func TestAIDLens(t *testing.T) {
	var file struct{ Cases []aidCase }
	load(t, "aid-lens.json", &file)
	for _, c := range file.Cases {
		t.Run(c.ID, func(t *testing.T) {
			key, err := FromSeed(mustHex(t, c.SeedHex))
			if err != nil {
				t.Fatal(err)
			}
			if key.AID() != c.AID {
				t.Errorf("AID = %q, want %q", key.AID(), c.AID)
			}
			if key.Keyid() != c.Keyid {
				t.Errorf("keyid = %q, want %q", key.Keyid(), c.Keyid)
			}
			public, err := VerifyingKey(c.AID)
			if err != nil {
				t.Fatal(err)
			}
			if hex.EncodeToString(public) != c.PublicKeyHex {
				t.Errorf("recovered key = %x, want %s", public, c.PublicKeyHex)
			}
		})
	}
}

func TestSignatureBaseVectors(t *testing.T) {
	var file struct{ Cases []baseCase }
	load(t, "signature-base.json", &file)
	for _, c := range file.Cases {
		t.Run(c.ID, func(t *testing.T) {
			base, err := SignatureBase(c.Method, c.URL, c.Headers, c.Covered,
				SignatureParams{Created: c.Created, Keyid: c.Keyid, Alg: c.Alg})
			if err != nil {
				t.Fatal(err)
			}
			if string(base) != c.Base {
				t.Errorf("base mismatch\n got: %q\nwant: %q", base, c.Base)
			}
			// Ed25519 is deterministic, so a port that builds the right base produces the right
			// bytes — this is byte equality, not a verification round trip.
			key, err := FromSeed(mustHex(t, c.SeedHex))
			if err != nil {
				t.Fatal(err)
			}
			got := base64.StdEncoding.EncodeToString(key.Sign(base))
			if got != c.Signature {
				t.Errorf("signature = %s, want %s", got, c.Signature)
			}
		})
	}
}

func TestAcceptVectors(t *testing.T) {
	var file struct{ Cases []requestCase }
	load(t, "accepts.json", &file)
	for _, c := range file.Cases {
		t.Run(c.ID, func(t *testing.T) {
			verdict, err := VerifyRequest(c.Method, c.URL, c.Headers, c.options())
			if err != nil {
				t.Fatalf("expected this request to verify, got %v", err)
			}
			if verdict.AID != c.AID {
				t.Errorf("AID = %q, want %q", verdict.AID, c.AID)
			}
			if !slices.Equal(verdict.Covered, c.Covered) {
				t.Errorf("covered = %v, want %v", verdict.Covered, c.Covered)
			}
		})
	}
}

func TestRefusalVectors(t *testing.T) {
	var file struct{ Cases []requestCase }
	load(t, "refusals.json", &file)
	for _, c := range file.Cases {
		t.Run(c.ID, func(t *testing.T) {
			// Every entry names the kind fiki reports, so this port maps its own onto the same
			// condition rather than inventing a taxonomy of its own.
			_, err := VerifyRequest(c.Method, c.URL, c.Headers, c.options())
			if err == nil {
				t.Fatalf("expected %s, got a verdict", c.Error)
			}
			var fikiErr *Error
			if !errors.As(err, &fikiErr) {
				t.Fatalf("expected a fiki.Error, got %T", err)
			}
			if fikiErr.Kind != c.Error {
				t.Errorf("kind = %s, want %s", fikiErr.Kind, c.Error)
			}
		})
	}
}

File: java/src/test/java/com/bakobo/fiki/VectorsTest.java
package com.bakobo.fiki;

import static org.junit.jupiter.api.Assertions.assertArrayEquals;
import static org.junit.jupiter.api.Assertions.assertEquals;
import static org.junit.jupiter.api.Assertions.assertThrows;
import static org.junit.jupiter.api.Assertions.assertTrue;

import com.fasterxml.jackson.databind.JsonNode;
import com.fasterxml.jackson.databind.ObjectMapper;
import java.io.File;
import java.nio.charset.StandardCharsets;
import java.util.ArrayList;
import java.util.Base64;
import java.util.HexFormat;
import java.util.LinkedHashMap;
import java.util.List;
import java.util.Map;
import java.util.stream.Stream;
import org.junit.jupiter.api.DynamicTest;
import org.junit.jupiter.api.TestFactory;

/**
 * The shared conformance vectors (this.i @5gf6r08f, @2tt6fmc0).
 *
 * <p>They live at the repository root rather than under java/ so this implementation and the other
 * four are held to the same bytes. A copy under each language is the drift the polyglot layout
 * exists to prevent, which is why this file reaches up rather than embedding anything.
 */
class VectorsTest {

    private static final ObjectMapper MAPPER = new ObjectMapper();

    private static JsonNode load(String name) throws Exception {
        File file = new File("../vectors/" + name);
        assertTrue(file.isFile(), "the shared vectors are not where every port reaches them: " + file);
        return MAPPER.readTree(file);
    }

    private static Map<String, String> headers(JsonNode node) {
        Map<String, String> out = new LinkedHashMap<>();
        if (node != null && node.isObject()) {
            node.fields().forEachRemaining(e -> out.put(e.getKey(), e.getValue().asText()));
        }
        return out;
    }

    private static List<String> strings(JsonNode node) {
        List<String> out = new ArrayList<>();
        node.forEach(item -> out.add(item.asText()));
        return out;
    }

    private static byte[] body(JsonNode node) {
        JsonNode body = node.get("body");
        return body == null || body.isNull() ? null : body.asText().getBytes(StandardCharsets.UTF_8);
    }

    private static Fiki.VerifyOptions options(JsonNode c) {
        JsonNode maxAge = c.get("max_age");
        Fiki.VerifyOptions opts = maxAge == null || maxAge.isNull()
            ? Fiki.VerifyOptions.decliningFreshness()
            : Fiki.VerifyOptions.maxAge(maxAge.asLong());
        opts = opts.withBody(body(c));
        JsonNode now = c.get("now");
        return now == null || now.isNull() ? opts : opts.withNow(now.asLong());
    }

    @TestFactory
    Stream<DynamicTest> aidLens() throws Exception {
        List<DynamicTest> tests = new ArrayList<>();
        for (JsonNode c : load("aid-lens.json").get("cases")) {
            tests.add(DynamicTest.dynamicTest(c.get("id").asText(), () -> {
                Key key = Key.fromSeed(HexFormat.of().parseHex(c.get("seed_hex").asText()));
                assertEquals(c.get("aid").asText(), key.aid());
                assertEquals(c.get("keyid").asText(), key.keyid());
                assertArrayEquals(
                    HexFormat.of().parseHex(c.get("public_key_hex").asText()),
                    Key.verifyingKeyBytes(c.get("aid").asText()));
            }));
        }
        return tests.stream();
    }

    @TestFactory
    Stream<DynamicTest> signatureBases() throws Exception {
        List<DynamicTest> tests = new ArrayList<>();
        for (JsonNode c : load("signature-base.json").get("cases")) {
            tests.add(DynamicTest.dynamicTest(c.get("id").asText(), () -> {
                JsonNode alg = c.get("alg");
                Fiki.Params params = new Fiki.Params(
                    c.get("created").asLong(), c.get("keyid").asText(),
                    alg == null ? null : alg.asText(), null, null, null);
                byte[] base = Fiki.signatureBase(
                    c.get("method").asText(), c.get("url").asText(), headers(c.get("headers")),
                    strings(c.get("covered")), params);
                assertEquals(c.get("base").asText(), new String(base, StandardCharsets.UTF_8));
                // Ed25519 is deterministic, so a port that builds the right base produces the
                // right bytes: byte equality, not a verification round trip.
                Key key = Key.fromSeed(HexFormat.of().parseHex(c.get("seed_hex").asText()));
                assertEquals(
                    c.get("signature").asText(),
                    Base64.getEncoder().encodeToString(key.sign(base)));
            }));
        }
        return tests.stream();
    }

    @TestFactory
    Stream<DynamicTest> accepts() throws Exception {
        List<DynamicTest> tests = new ArrayList<>();
        for (JsonNode c : load("accepts.json").get("cases")) {
            tests.add(DynamicTest.dynamicTest(c.get("id").asText(), () -> {
                Fiki.Verdict verdict = Fiki.verifyRequest(
                    c.get("method").asText(), c.get("url").asText(), headers(c.get("headers")), options(c));
                assertEquals(c.get("aid").asText(), verdict.aid());
                assertEquals(strings(c.get("covered")), verdict.covered());
            }));
        }
        return tests.stream();
    }

    @TestFactory
    Stream<DynamicTest> refusals() throws Exception {
        List<DynamicTest> tests = new ArrayList<>();
        for (JsonNode c : load("refusals.json").get("cases")) {
            tests.add(DynamicTest.dynamicTest(c.get("id").asText(), () -> {
                // Every entry names the kind fiki reports, so this port maps its own onto the same
                // condition rather than inventing a taxonomy of its own.
                FikiException thrown = assertThrows(FikiException.class, () -> Fiki.verifyRequest(
                    c.get("method").asText(), c.get("url").asText(), headers(c.get("headers")), options(c)));
                assertEquals(c.get("error").asText(), thrown.kind().name());
            }));
        }
        return tests.stream();
    }
}
